### PR TITLE
async point counts

### PR DESCRIPTION
**Pull request by emre2038**
## Description
* Getting the point counts async after the facilities are created. 
   - Facilities are created with a loading spinner instead of facility count 
   - The facility count API is called(this uses dummy data for now until the backend is ready)
   - Facility counts are populated and the loading spinners are removed

* Loading state while the facilities are getting created
   - When a new geo is clicked, facilities go into the loading state immediately
   - `Loading` text is shown instead of facilities description
   - `Loading` button is shown instead of Show/Hide facilities button
   - `Download all data` button is disabled
   - Facilities pane is hidden
   - This is how it looks when it goes into the loading state
image

* Refactored the facilities code too so the code is simpler to read. Before we used `src/js/profile/profile_header.js` to create the facilities but now there is `facilities` folder in `profile` folder. 

image

In the image above
a - `src/js/profile/profile_header.js` manages breadcrumbs and geo description
b - `src/js/profile/facilities/facility_controller.js` manages the facility wrapper, facilities description, downloading all data, expanding / hiding the facilities etc
c - `src/js/profile/facilities/facility.js` manages the facility cards
d - `src/js/profile/facilities/facilityItem.js` manages each row of facility cards

## Related Issue
https://wazimap.atlassian.net/browse/WNCM-19

## How is it tested automatically?
* A GUI test is created `__tests__/gui/facilities.feature`
* Tests if the facilities are created correctly
* Tests if it goes into the loading state when the map is navigating to a new geo

## How should a reviewer test it locally
* Expand the rich data panel
* Confirm that the facilities are created correctly
* Confirm that the downloads are working
* Confirm that when navigating to a new geo, facilities pane goes into the loading state correctly

### Added
* `src/js/profile/facilities/facility_controller.js` manages the facility wrapper, facilities description, downloading all data, expanding / hiding the facilities etc
* `src/js/profile/facilities/facility.js` manages the facility cards
* `src/js/profile/facilities/facilityItem.js` manages each row of facility cards
*  `__tests__/gui/facilities.feature` GUI test

### Updated
* `src/js/profile/profile_header.js` removed the code that is related with facilities

## Checklist

- [x]  🚀 is the code ready to be merged and go live?
- [x]  🛠 does it work (build) locally
- [x] 👩‍🎨 does the design match the [Demo](https://wazimap-ng-v1.webflow.io/demo)

### Pull Request

- [x]  📰 good title
- [x]  📝good description
- [x]  🔖 issue linked
- [x]  📖 changelog filled out
- [x] commit messages are meaningful

### Code Quality

- [x]  🚧 no commented out code
- [x]  🖨 no unnecessary logging
- [x]  🎱 no magic numbers

### Testing

- [x]  ✅ added (appropriate) unit tests
- [x]  💢 edge cases in tests were considered
- [x]  ✅ ran tests locally & are passing

### CODE REVIEW

File: __tests__/gui/facilities/facilities.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Then('I wait until map is ready', () => {
+    waitUntilGeographyIsLoaded('South Africa Test');
+})
+
+When('I expand Rich Data Panel', () => {
+    expandRichDataPanel();
+})
+
+Then('I click on Show Locations button', () => {
+    cy.wait(500);
```
[the remark this answers is not in the document]
Comment (jbothma) [reply __tests__/gui/facilities/facilities.js:36]: what are we waiting for?

File: __tests__/gui/facilities/facilities.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+When('I check if the location count is correct', () => {
+    cy.get('.rich-data .location__facilities_categories-value strong').should('have.text', '4');
+})
+
+Then('I check if the facilities are created correctly', () => {
+    const themeCount = themes_count.length;
+    cy.get('.rich-data .location__facilities_content-wrapper .location-facility').should('have.length', themeCount);
+    cy.get('.rich-data .location__facilities_content-wrapper .location-facility').each(($div, index) => {
+        const themeTitle = themes_count[index].name
```
Comment (jbothma) [inline __tests__/gui/facilities/facilities.js:48]: I prefer hardcoding assertion values like this in the test, rather than pulling it from the test data dynamically. Otherwise I worry that someone changes the test data to something defaulty like 0 in future without realising its relevance to the test, and then breaking the test later without realising 

You don't have to change this, just sharing my preference for thought/discussion

File: __tests__/gui/facilities/facilities.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    const firstCategory = themes_count[0].subthemes[0].label;
+    cy.get('.rich-data .location__facilities_content-wrapper .location-facility .location-facility__item_name div').first().should('have.text', firstCategory);
+})
+
+When('I navigate to EC and check if the loading state is displayed correctly', () => {
+    let sendResponse;
+    const trigger = new Promise((resolve) => {
+        sendResponse = resolve;
+    });
```
Comment (jbothma) [inline __tests__/gui/facilities/facilities.js:62]: :clap: :clap: :clap: 
https://blog.dai.codes/cypress-loading-state-tests/

**Comment by jbothma** [comment]
It doesn't look like GUI tests are running in this PR - any idea why?

File: __tests__/gui/facilities/facilities.js (hunk 3 above)
Comment (emre2038) [reply __tests__/gui/facilities/facilities.js:62]: sure - this was the link I found this pattern btw

File: __tests__/gui/facilities/facilities.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (emre2038) [reply __tests__/gui/facilities/facilities.js:36]: removed waiting - added timeout instead

File: src/js/profile/facilities/facility_controller.js [hunk 4]
```diff
@@ -84,11 +84,13 @@ export class FacilityController extends Component {
     }
 
     getAndAddFacilities() {
-        this.model.api.getThemesCount(this.model.profileId, this.model.geography.code)
-            .then((data) => {
-                this.model.themes = data;
-                this.addFacilities();
-            })
+        if(this.model.api !== null){
```
Comment (jbothma) [inline src/js/profile/facilities/facility_controller.js:87]: when would this be null?

File: __tests__/profile/profile_header.test.js [hunk 5]
```diff
@@ -178,21 +163,27 @@ describe('downloading facilities', () => {
...
+        let f = new FacilityItem(component, facilityRowClone, category, true, false);
+        const downloadBtn = $(f.facilityItem).find('.location-facility__item_download')[0];
+
+        expect(downloadBtn).not.toHaveClass('hidden');
     })
 
     test('download is disabled', () => {
-        config = {
-            richdata: {
-                hide: ['facility-downloads']
```
Comment (jbothma) [inline __tests__/profile/profile_header.test.js:188]: why are we losing this config? I think it's broken because I couldn't hide the button with this config recently but the test still states that it's testing the hidden functionality.

How does this test even pass if it's the same as the one above but asserting the opposite?

File: src/js/profile/facilities/facility_controller.js (hunk 4 above)
Comment (emre2038) [reply src/js/profile/facilities/facility_controller.js:87]: jest throws exception because it does not recognize `fetch` and I didn't want to import `jest-fetch-mock` just for this so I just send `null` so the api is not called - what do you think about this workaround?
https://github.com/OpenUpSA/wazimap-ng-ui/blob/741ae253e7b82a98a8ddf8382c402a59ca640e68/__tests__/profile/profile_header.test.js#L129

File: __tests__/profile/profile_header.test.js (hunk 5 above)
Comment (emre2038) [reply __tests__/profile/profile_header.test.js:188]: the test above passes `false` for `isDownloadsDisabled` parameter and expects `downloadBtn` to not be hidden
https://github.com/OpenUpSA/wazimap-ng-ui/blob/741ae253e7b82a98a8ddf8382c402a59ca640e68/__tests__/profile/profile_header.test.js#L172 

while the test below passes `true` and expects `downloadBtn` to be hidden
https://github.com/OpenUpSA/wazimap-ng-ui/blob/741ae253e7b82a98a8ddf8382c402a59ca640e68/__tests__/profile/profile_header.test.js#L185

Comment (emre2038) [reply __tests__/profile/profile_header.test.js:188]: let me double check the config again. but here I test the FacilityItem object while the `config` is processed in `ProfileHeader`. that's why I removed `config` but let me add another test to confirm that the `config` element is processed correctly

Comment (emre2038) [reply __tests__/profile/profile_header.test.js:188]: added back the config

Comment (jbothma) [reply __tests__/profile/profile_header.test.js:188]: sounds like the config is irrelevant to the test and in the full codebase we rely on some code above new FacilityItem() to convert the config into the right parameter to FacilityItem constructor, right?
